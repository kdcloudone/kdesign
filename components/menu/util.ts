import isArray from 'lodash/isArray'
import * as React from 'react'
import { toArray } from '../_utils/react-children'

const DEFAULT_ITEM_WIDTH = 160
const DEFAULT_MAX_WIDTH = 800
const DEFAULT_MARGIN_WIDTH = 40
const DEFAULT_ITEM_NUM = 5

export const isString = (value: any): boolean => {
  return typeof value === 'string'
}

export const isNumber = (value: any): boolean => {
  return typeof value === 'number'
}

export const getColumnWidthList = (nodeList: React.ReactElement): number[] => {
  if (!nodeList) return []

  const columnWidthList: number[] = []

  toArray(nodeList).map((item: React.ReactElement) => {
    if (!item || isString(item) || isNumber(item)) return item

    const width = getItemWidth(item.props.children)
    columnWidthList.push(width)

    return item
  })

  return columnWidthList
}

// 计算外层面板的总宽度 DEFAULT_MAX_WIDTH: 824px, 最多四列（790px），
// 每列160px+40px
// 其中160px是width,40px是marginRight
export const getWrapWidth = (nodelist: React.ReactElement): number => {
  let wrapWidth = 0
  let totalWidth = 0

  const columnWidthList = getColumnWidthList(nodelist)

  columnWidthList.map((width: number) => {
    totalWidth += width + DEFAULT_MARGIN_WIDTH

    if (totalWidth < DEFAULT_MAX_WIDTH) {
      wrapWidth = totalWidth
    } else {
      wrapWidth = DEFAULT_MAX_WIDTH
    }

    return width
  })
  return wrapWidth + 10
}

// 计算子项的宽度
export const getItemWidth = (children: React.ReactElement): number => {
  if (!isArray(children)) {
    return DEFAULT_ITEM_WIDTH
  }

  if (!children || children.length <= DEFAULT_ITEM_NUM) {
    return DEFAULT_ITEM_WIDTH
  } else {
    const width: number =
      (DEFAULT_ITEM_WIDTH + DEFAULT_MARGIN_WIDTH) * Math.ceil(children.length / DEFAULT_ITEM_NUM) - DEFAULT_MARGIN_WIDTH
    return width > DEFAULT_MAX_WIDTH ? DEFAULT_MAX_WIDTH : width
  }
}

export const renderReactNodeFunction = (reactNode: React.ReactNode) => {
  if (typeof reactNode === 'function') {
    return reactNode()
  }
  return reactNode
}

export const isElementInViewport = (el: any) => {
  const rect = el.getBoundingClientRect()
  return (
    rect.top >= 0 &&
    rect.left >= 0 &&
    rect.bottom <= (window.innerHeight || document.documentElement.clientHeight) &&
    rect.right <= (window.innerWidth || document.documentElement.clientWidth)
  )
}

export const DEFAUTL_PADDING = 15
