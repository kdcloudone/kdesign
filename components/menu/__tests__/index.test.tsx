import React from 'react'
import { mount, render } from 'enzyme'
import Menu from '../index'
import Icon from '../../icon'
import mountTest from '../../../tests/shared/mountTest'
import { isString, isNumber, getColumnWidthList, getWrapWidth, getItemWidth } from '../util'

const MenuModes = ['inline', 'vertical']

describe('Menu', () => {
  // 1.mount test
  mountTest(Menu)

  mountTest(() => <Menu mode={'inline'} />)
  mountTest(() => <Menu mode={'vertical'} />)

  // 2.render test
  it('renders correctly', () => {
    expect(
      render(
        <Menu>
          <Menu.Item key="1">222</Menu.Item>
        </Menu>,
      ),
    ).toMatchSnapshot()
  })

  // 4. render null or undefined without errors
  it('render null or undefined buttons without errors', () => {
    const wrapper = mount(
      <Menu>
        {null}
        {undefined}
      </Menu>,
    )
    expect(wrapper.find('li').length).toBe(0)
  })

  // 5. displayName
  it('should have displayName static property', () => {
    const wrapper = mount(<Menu></Menu>)
    expect((wrapper.type() as any).displayName).toBe('Menu')
  })

  // 6. class state
  it('should class use right', () => {
    // default
    const defaultMenu = mount(<Menu />)
    expect(defaultMenu.find('.kd-menu')).toHaveClassName('kd-menu-vertical')
  })

  // 7. mouseEvent
  it('should mouseEvent use right', () => {
    const props: any = {
      onMouseLeave: jest.fn(),
      onMouseEnter: jest.fn(),
      onClick: jest.fn(),
      triggerSubMenuAction: 'click',
    }

    const itemProps: any = {
      onMouseLeave: jest.fn(),
      onMouseEnter: jest.fn(),
    }

    // default
    const menuWrapper = mount(
      <Menu {...props}>
        <Menu.Item {...itemProps}>111</Menu.Item>
      </Menu>,
    )
    menuWrapper.simulate('mouseleave')
    menuWrapper.simulate('mouseenter')

    expect(props.onMouseLeave).toHaveBeenCalled()
    expect(props.onMouseEnter).toHaveBeenCalled()

    props.onMouseLeave = ''
    props.onMouseEnter = ''
    props.selectedKey = 1
    props.openKeys = ['sub1']
    menuWrapper.setProps(props)

    menuWrapper.find('li').simulate('click')

    expect(props.onClick).toHaveBeenCalled()

    menuWrapper.find('li').simulate('mouseleave')
    menuWrapper.find('li').simulate('mouseenter')

    expect(itemProps.onMouseLeave).toHaveBeenCalled()
    expect(itemProps.onMouseEnter).toHaveBeenCalled()

    props.triggerSubMenuAction = 'hover'

    const subMenuWrapper = mount(
      <Menu {...props}>
        <Menu.SubMenu title="sub1" icon={<Icon type="add" />} popupOffset={[20]}>
          <Menu.Item>1111</Menu.Item>
        </Menu.SubMenu>
        <Menu.SubMenu title="sub2" subMenuMode="horizontal" popupOffset={[]}>
          <Menu.Item>1111</Menu.Item>
        </Menu.SubMenu>
      </Menu>,
    )

    subMenuWrapper.find('.kd-menu-submenu').at(0).simulate('click')
    subMenuWrapper.find('.kd-menu-submenu').at(0).simulate('mouseenter')
    subMenuWrapper.find('.kd-menu-submenu').at(0).simulate('mouseleave')

    props.mode = 'vertical'
    props.triggerSubMenuAction = 'click'
    props.collapsed = true
    subMenuWrapper.update()

    props.mode = 'inline'
    props.triggerSubMenuAction = 'click'
    props.collapsed = false
    subMenuWrapper.update()

    MenuModes.forEach((mode) => {
      const testMenu = mount(
        <Menu mode={mode} {...props}>
          <Menu.SubMenu title="sub1">
            <Menu.Item>1111</Menu.Item>
            <Menu.Item>{1111}</Menu.Item>
          </Menu.SubMenu>
        </Menu>,
      )
      testMenu.find('.kd-menu-submenu').simulate('click')
    })
  })

  it('other status', () => {
    const props = {
      collapsed: true,
    }

    const wrapperItem = mount(
      <Menu {...props}>
        <Menu.Item disabled>1111</Menu.Item>
        <Menu.SubMenu title="sub1">
          <Menu.Item>1111</Menu.Item>
          <Menu.Item>{1111}</Menu.Item>
        </Menu.SubMenu>
        <Menu.Item icon={<Icon type="add" />}>1111</Menu.Item>
      </Menu>,
    )
    wrapperItem.find('.kd-menu-item').at(0).simulate('click')
    wrapperItem.find('.kd-menu-item').at(0).simulate('mouseenter')
    wrapperItem.find('.kd-menu-item').at(0).simulate('mouseleave')

    // const wrapperItem = mount(
    //   <Menu {...props} triggerSubMenuAction="click">
    //     <Menu.SubMenu title="sub1" icon={<Icon type="add" />}>
    //       <Menu.Item>1111</Menu.Item>
    //       <Menu.SubMenu title="sub2">
    //         <Menu.Item>1111</Menu.Item>
    //       </Menu.SubMenu>
    //     </Menu.SubMenu>
    //   </Menu>,
    // )
  })

  it('util test', () => {
    expect(isString('222')).toBe(true)
    expect(isString(222)).toBe(false)

    expect(isNumber('222')).toBe(false)
    expect(isNumber(222)).toBe(true)

    expect(getColumnWidthList(<div />)).toEqual([160])
    expect(getColumnWidthList(<></>)).toEqual([160])

    expect(getWrapWidth(<div />)).toBe(210)

    expect(getItemWidth(<div />)).toBe(160)
  })
})
