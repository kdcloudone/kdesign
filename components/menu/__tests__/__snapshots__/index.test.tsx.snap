// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly 1`] = `
<div
  class="kd-menu kd-menu-vertical kd-menu-collapsed kd-menu-dark"
  role="menu"
  style="transition:width 0.2s cubic-bezier(0,.4,.4,1)"
>
  <ul>
    <li
      class="kd-menu-item"
      role="menuitem"
    >
      <div
        class="kd-menu-item-title"
        style="padding-left:50px;padding-right:0"
      >
        <span>
          222
        </span>
      </div>
    </li>
  </ul>
</div>
`;
